### PR TITLE
K8SPSMDB-198: Failover update

### PR DESCRIPTION
**Pull request by blez**
[K8SPSMDB-198](https://jira.percona.com/browse/K8SPSMDB-198)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | failed |
| upgrade | failed |
| init-deploy | failed |
| monitoring | passed |
| limits | failed |
| self-healing | failed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| operator-self-healing | failed |
| security-context | passed |
| scheduled-backup | passed |
| arbiter | failed |
| service-per-pod | failed |
| liveness | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | failed |
| storage | passed |
| init-deploy | failed |
| upgrade | failed |
| limits | failed |
| self-healing | passed |
| monitoring | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| monitoring-2-0 | failed |
| security-context | passed |
| operator-self-healing | failed |
| arbiter | failed |
| scheduled-backup | failed |
| service-per-pod | failed |
| liveness | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | failed |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| upgrade | failed |
| init-deploy | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | failed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| arbiter | failed |
| service-per-pod | failed |
| liveness | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | failed |
| upgrade | passed |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| liveness | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-368-f584a284

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| one-pod | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | failed |
| upgrade | passed |
| monitoring | passed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| security-context | passed |
| arbiter | passed |
| operator-self-healing | failed |
| scheduled-backup | failed |
| service-per-pod | passed |
| liveness | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade | failed |
| upgrade-consistency | passed |
| monitoring | failed |
| self-healing | failed |
| limits | passed |
| demand-backup | passed |
| monitoring-2-0 | failed |
| scaling | passed |
| operator-self-healing | failed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| service-per-pod | passed |
| liveness | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| storage | passed |
| self-healing | failed |
| one-pod | passed |
| monitoring | failed |
| upgrade | passed |
| monitoring-2-0 | failed |
| operator-self-healing | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| storage | passed |
| self-healing | failed |
| one-pod | passed |
| monitoring | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| monitoring-2-0 | failed |
| operator-self-healing | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-368-bf250147

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| one-pod | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| storage | passed |
| self-healing | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| operator-self-healing | passed |

File: pkg/controller/perconaservermongodb/smart.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+		if k8sErrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, errors.Wrap(err, "get backup list")
+	}
+
+	for _, bcp := range bcps.Items {
+		if bcp.Status.State == api.BackupStateRunning ||
+			bcp.Status.State == api.BackupStateWaiting ||
+			bcp.Status.State == api.BackupStateRequested {
```
Comment (delgod) [inline pkg/controller/perconaservermongodb/smart.go:114]: can we check status via successful or failed state?
reason - we can add more statuses in future 
```suggestion
		if b.Status.State != api.BackupStateReady && b.Status.State != api.BackupStateError && {
```

File: pkg/controller/perconaservermongodb/smart.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+			primaryPod = pod
+		} else {
+			log.Info(fmt.Sprintf("apply changes to secondary pod %s", pod.Name))
+			if err := r.applyNWait(cr, sfs.Status.UpdateRevision, &pod, waitLimit); err != nil {
+				return fmt.Errorf("failed to apply changes: %v", err)
+			}
+		}
+	}
+
+	log.Info(fmt.Sprintf("apply changes to primary pod %s", primaryPod.Name))
```
Comment (delgod) [inline pkg/controller/perconaservermongodb/smart.go:94]: can we do primary stepdown before deleting Pod?

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | failed |
| storage | failed |
| init-deploy | failed |
| upgrade | passed |
| limits | passed |
| monitoring | failed |
| self-healing | failed |
| scaling | passed |
| monitoring-2-0 | passed |
| upgrade-consistency | passed |
| operator-self-healing | failed |
| security-context | failed |
| demand-backup | failed |
| arbiter | failed |
| scheduled-backup | failed |
| service-per-pod | passed |
| liveness | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | failed |
| storage | failed |
| upgrade | failed |
| monitoring | passed |
| upgrade-consistency | passed |
| init-deploy | passed |
| demand-backup | failed |
| self-healing | passed |
| scheduled-backup | failed |
| limits | passed |
| scaling | failed |
| security-context | passed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| arbiter | failed |
| service-per-pod | failed |
| liveness | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | failed |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| one-pod | passed |
| self-healing | passed |
| operator-self-healing | passed |
| upgrade | passed |
| security-context | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| service-per-pod | passed |
| scheduled-backup | failed |
| liveness | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| storage | passed |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | failed |
| self-healing | passed |
| operator-self-healing | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade | failed |
| init-deploy | failed |
| storage | failed |
| one-pod | failed |
| self-healing | failed |
| limits | passed |
| monitoring | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | failed |
| operator-self-healing | passed |
| security-context | passed |
| monitoring-2-0 | failed |
| scheduled-backup | failed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| monitoring | passed |
| monitoring-2-0 | failed |
| upgrade-consistency | passed |
| demand-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| one-pod | passed |
| monitoring | passed |
| monitoring-2-0 | failed |
| upgrade | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| monitoring | passed |
| monitoring-2-0 | failed |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| upgrade | passed |
| one-pod | passed |
| self-healing | failed |
| monitoring | failed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | failed |
| monitoring-2-0 | failed |
| demand-backup | failed |
| operator-self-healing | passed |
| arbiter | failed |
| security-context | passed |
| scheduled-backup | passed |
| service-per-pod | failed |
| liveness | failed |

**Pull request merged**